src/App.js

```diff
@@ -522,7 +526,31 @@ export default class App extends React.Component<{}, AppState> {
         );
     }
 
+    componentDidMount() {
+        if (window.location.search != null) {
+            const query = new URLSearchParams(window.location.search);
+            const programQuery = query.get('program');
+            if (programQuery != null) {
+                try {
+                    this.setState({
+                        program: this.programSerializer.deserialize(programQuery)
+                    });
+                } catch(err) {
+                    console.log(`Error parsing program: ${programQuery}`);
+                    console.log(err.toString());
+                }
+            }
+        }
+    }
+
     componentDidUpdate(prevProps: {}, prevState: AppState) {
+        if (this.state.program !== prevState.program) {
+            const encodedURI = encodeURI(this.programSerializer.serialize(this.state.program));
+            window.history.pushState(
+                {program: encodedURI},
+                '',
+                `?program=${encodedURI}`);
```

Comment:
Let's make a function for generating the URL that takes an unencoded version string and and unencoded program string and does the necessary encoding and build the URL.